### PR TITLE
[libSwiftScan][Caching] Add async download API

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Add a new async download API from CASID. This helps build system better schedule the job/remove duplicated downloads, than using an opaque cached_output handle.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
ping

File: tools/libSwiftScan/SwiftCaching.cpp [hunk 1]
```diff
@@ -467,6 +467,98 @@ bool swiftscan_cached_output_load(swiftscan_cached_output_t output,
   return success();
 }
 
+namespace {
+/// Asynchronously visits the graph of the object node to ensure it's fully
+/// materialized.
+class AsyncObjectLoader
+    : public std::enable_shared_from_this<AsyncObjectLoader> {
```
Comment (artemcm) [inline tools/libSwiftScan/SwiftCaching.cpp:474]: I didn't know about this, neat.

**Review by artemcm — APPROVED** [review]

**Comment by artemcm** [comment]
Will we test this from the `swift-driver` repo?

**Comment by cachemeifyoucan** [comment]
Test for new API added here: https://github.com/apple/swift-driver/pull/1491

**Pull request merged**